client/src/extension.ts

```diff
@@ -126,8 +127,10 @@ export async function activate (context: vscode.ExtensionContext): Promise<void>
 
   taskProvider = vscode.tasks.registerTaskProvider('bitbake', bitbakeTaskProvider)
 
+  context.subscriptions.push(bitbakeRecipeScanner)

```

Comment:
I don't think the code is very clear about what's the event and the action. bitbakeRecipeScanner acts as an uneeded middleware IMO.

Reply:
I do not understand when is ServerScanRequestListener fired and it doesn't look like it uses bitbakeRecipeScanner's data in any way.

Reply:
The event emitter is `serverScanRequest`. It is fired in subscribeToTaskEnd when the scan is done. Then all the listeners are called. This is just the [observer pattern](https://en.wikipedia.org/wiki/Observer_pattern).

Do you have an other idea how `reviewDiagnostics` could be called once the scan is finished?

Reply:
It looks like an observer pattern on an observer pattern with a superfluous layer. I've pushed a commit on top to illustrate a simpler code.